File: frontend/src/api/careerProcess.ts
import { apiClient } from './client';
import type {
  JobDescriptionSnapshot,
  JobOpportunity,
  JobOpportunityMerge,
  JobOpportunityMergeCandidate,
  NextAction,
  NextActionStatus,
  NextActionTimeKind,
  ProcessEvent,
  ProcessEventKind,
} from '@/types/career';

export interface OpportunityCreateInput {
  company_name: string;
  job_title: string;
  entry_reason: 'explicit_tracking' | 'targeted_preparation' | 'user_confirmed_application';
  occurred_at: string;
  source_kind: 'user_assertion';
  source_identity: string;
  source_description: string;
  location?: string;
  team?: string;
  source_url?: string;
  idempotency_key?: string;
  directions?: Array<{ direction_id: string; match_reason: string }>;
}

export async function listJobOpportunities(includeArchived = false): Promise<JobOpportunity[]> {
  return (
    await apiClient.get('/career-process/opportunities', {
      params: { include_archived: includeArchived },
    })
  ).data;
}

export async function createJobOpportunity(input: OpportunityCreateInput): Promise<JobOpportunity> {
  return (await apiClient.post('/career-process/opportunities', input)).data;
}

export async function listJobOpportunityMergeCandidates(): Promise<JobOpportunityMergeCandidate[]> {
  return (await apiClient.get('/career-process/opportunity-merge-candidates')).data;
}

export async function listJobOpportunityMerges(includeRetracted = false): Promise<JobOpportunityMerge[]> {
  return (
    await apiClient.get('/career-process/opportunity-merges', {
      params: { include_retracted: includeRetracted },
    })
  ).data;
}

export async function mergeJobOpportunities(input: {
  duplicateOpportunityId: string;
  canonicalOpportunityId: string;
  reason: string;
  operationKey: string;
}): Promise<JobOpportunityMerge> {
  return (
    await apiClient.post('/career-process/opportunity-merges', {
      duplicate_opportunity_id: input.duplicateOpportunityId,
      canonical_opportunity_id: input.canonicalOpportunityId,
      operation_key: input.operationKey,
      reason: input.reason,
    })
  ).data;
}

export async function retractJobOpportunityMerge(input: {
  merge: JobOpportunityMerge;
  reason: string;
  operationKey: string;
}): Promise<JobOpportunityMerge> {
  return (
    await apiClient.post(
      `/career-process/opportunity-merges/${encodeURIComponent(input.merge.id)}/retract`,
      {
        expected_version: input.merge.version,
        operation_key: input.operationKey,
        reason: input.reason,
      },
    )
  ).data;
}

export async function replaceJobOpportunityDirections(input: {
  opportunityId: string;
  expectedVersion: number;
  sourceIdentity: string;
  directions: Array<{ direction_id: string; match_reason: string }>;
}): Promise<JobOpportunity> {
  return (
    await apiClient.patch(
      `/career-process/opportunities/${encodeURIComponent(input.opportunityId)}/directions`,
      {
        expected_version: input.expectedVersion,
        source_kind: 'user_assertion',
        source_identity: input.sourceIdentity,
        directions: input.directions,
      },
    )
  ).data;
}

export async function listProcessEvents(opportunityId: string): Promise<ProcessEvent[]> {
  return (
    await apiClient.get(`/career-process/opportunities/${encodeURIComponent(opportunityId)}/events`)
  ).data;
}

export async function appendProcessEvent(input: {
  opportunityId: string;
  kind: ProcessEventKind;
  occurredAt: string;
  description: string;
  stepSummary?: string;
  applicationChannel?: string;
  sourceIdentity: string;
  idempotencyKey: string;
}): Promise<ProcessEvent> {
  return (
    await apiClient.post(
      `/career-process/opportunities/${encodeURIComponent(input.opportunityId)}/events`,
      {
        kind: input.kind,
        occurred_at: input.occurredAt,
        source_kind: 'user_assertion',
        source_identity: input.sourceIdentity,
        description: input.description,
        step_summary: input.stepSummary || undefined,
        application_channel: input.applicationChannel || undefined,
        idempotency_key: input.idempotencyKey,
      },
    )
  ).data;
}

export async function correctProcessEvent(input: {
  opportunityId: string;
  eventId: string;
  kind: ProcessEventKind;
  occurredAt: string;
  description: string;
  stepSummary?: string;
  applicationChannel?: string;
  sourceIdentity: string;
  idempotencyKey: string;
}): Promise<ProcessEvent> {
  return (
    await apiClient.post(
      `/career-process/opportunities/${encodeURIComponent(input.opportunityId)}/events/${encodeURIComponent(input.eventId)}/corrections`,
      {
        replacement_kind: input.kind,
        occurred_at: input.occurredAt,
        source_kind: 'user_assertion',
        source_identity: input.sourceIdentity,
        description: input.description,
        step_summary: input.stepSummary || undefined,
        application_channel: input.applicationChannel || undefined,
        idempotency_key: input.idempotencyKey,
      },
    )
  ).data;
}

export async function listNextActions(statuses?: NextActionStatus[]): Promise<NextAction[]> {
  return (await apiClient.get('/career-process/next-actions', { params: { statuses } })).data;
}

export interface NextActionCreateInput {
  content: string;
  status: 'suggested' | 'planned';
  time_kind: NextActionTimeKind;
  source_kind: 'user_request';
  source_identity: string;
  job_opportunity_id?: string;
  starts_at?: string;
  ends_at?: string;
  due_at?: string;
  original_time_text?: string;
  source_timezone?: string;
  interview_record_id?: string;
  offer_id?: string;
  artifact_id?: string;
  reminder_at?: string;
  reminder_channel?: 'in_app';
  idempotency_key: string;
}

export async function createNextAction(input: NextActionCreateInput): Promise<NextAction> {
  return (await apiClient.post('/career-process/next-actions', input)).data;
}

export async function editNextAction(input: {
  action: NextAction;
  content: string;
  time_kind: NextActionTimeKind;
  job_opportunity_id?: string;
  interview_record_id?: string;
  offer_id?: string;
  artifact_id?: string;
  starts_at?: string;
  ends_at?: string;
  due_at?: string;
  original_time_text?: string;
  source_timezone?: string;
  reminder_at?: string;
  reminder_channel?: 'in_app';
}): Promise<NextAction> {
  return (
    await apiClient.put(`/career-process/next-actions/${encodeURIComponent(input.action.id)}`, {
      expected_version: input.action.version,
      content: input.content,
      time_kind: input.time_kind,
      job_opportunity_id: input.job_opportunity_id,
      interview_record_id: input.interview_record_id,
      offer_id: input.offer_id,
      artifact_id: input.artifact_id,
      starts_at: input.starts_at,
      ends_at: input.ends_at,
      due_at: input.due_at,
      original_time_text: input.original_time_text,
      source_timezone: input.source_timezone,
      reminder_at: input.reminder_at,
      reminder_channel: input.reminder_channel,
    })
  ).data;
}

export async function transitionNextAction(
  action: NextAction,
  transition: 'plan' | 'complete' | 'close',
  reason?: string,
): Promise<NextAction> {
  return (
    await apiClient.post(`/career-process/next-actions/${encodeURIComponent(action.id)}/${transition}`, {
      source_kind: 'user_assertion',
      source_identity: `ui:${crypto.randomUUID()}`,
      ...(transition === 'close' ? { reason: reason || '用户关闭' } : {}),
    })
  ).data;
}

export async function listJobDescriptionSnapshots(
  opportunityId: string,
): Promise<JobDescriptionSnapshot[]> {
  return (
    await apiClient.get(
      `/career-process/opportunities/${encodeURIComponent(opportunityId)}/jd-snapshots`,
    )
  ).data;
}

export async function createJobDescriptionSnapshot(
  opportunityId: string,
  input: {
    source_kind: 'typed_product_ui';
    source_identity: string;
    source_version: string;
    original_url: string;
    observed_at: string;
    provider: string;
    canonical_content: string;
    idempotency_key: string;
  },
): Promise<JobDescriptionSnapshot> {
  return (
    await apiClient.post(
      `/career-process/opportunities/${encodeURIComponent(opportunityId)}/jd-snapshots`,
      input,
    )
  ).data;
}

File: backend/app/api/career_process.py
"""Authenticated HTTP boundary for the Stage 2 career-process slice.

The router is intentionally thin: product writes go through the same
``career_process_service`` commands used by future Agent tools and connector
handlers.  This public boundary can author user assertions and can reference
an owned, active ProcessEvent.  It cannot manufacture Observation, ToolResult,
receipt, Agent suggestion, or CopilotPreference identities; those source kinds
must enter through the concrete trusted handler that owns and validates them.
"""

from __future__ import annotations

from collections.abc import Callable
from typing import TypeVar

from fastapi import APIRouter, Depends, HTTPException, Query, Response, status
from sqlalchemy.orm import Session

from app.core.security import get_current_user
from app.db.database import get_db
from app.models.user import User
from app.schemas.job_opportunity import (
    JobOpportunityView,
    NextActionClose,
    NextActionCreate,
    NextActionEdit,
    NextActionStatus,
    NextActionTransition,
    NextActionView,
    OpportunityCreate,
    OpportunityDirectionsReplace,
    OpportunityMergeCandidateView,
    OpportunityMergeCreate,
    OpportunityMergeRetract,
    OpportunityMergeView,
    ProcessEventAppend,
    ProcessEventCorrection,
    ProcessEventView,
)
from app.schemas.job_description_snapshot import (
    JobDescriptionSnapshotFromProductUI,
    JobDescriptionSnapshotView,
)
from app.services.job_description_snapshot_service import (
    JobDescriptionSnapshotError,
    create_job_description_snapshot,
    list_job_description_snapshots,
)
from app.services.career_process_service import (
    CareerIdempotencyConflictError,
    CareerObjectNotFoundError,
    CareerProcessError,
    NextActionTransitionError,
    OpportunityArchivedError,
    OpportunityDirectionConflictError,
    OpportunityMergeConflictError,
    ProcessEventConflictError,
    append_confirmed_process_event,
    close_next_action,
    complete_next_action,
    correct_process_event,
    create_job_opportunity,
    create_next_action,
    edit_next_action,
    list_job_opportunities,
    list_opportunity_merges,
    list_next_actions,
    list_process_events,
    plan_next_action,
    merge_job_opportunities,
    retract_job_opportunity_merge,
    replace_job_opportunity_directions,
    suggest_opportunity_merge_candidates,
)


router = APIRouter(prefix="/career-process", tags=["career-process"])

_T = TypeVar("_T")


def _domain_http_error(exc: CareerProcessError) -> HTTPException:
    if isinstance(exc, CareerObjectNotFoundError):
        return HTTPException(status_code=404, detail="Career process object not found")
    if isinstance(
        exc,
        (
            CareerIdempotencyConflictError,
            NextActionTransitionError,
            OpportunityArchivedError,
            OpportunityDirectionConflictError,
            OpportunityMergeConflictError,
            ProcessEventConflictError,
        ),
    ):
        return HTTPException(status_code=409, detail=str(exc))
    return HTTPException(status_code=422, detail=str(exc))


def _run_domain(
    db: Session,
    operation: Callable[[], _T],
    *,
    commit: bool = False,
) -> _T:
    try:
        result = operation()
        if commit:
            db.commit()
        return result
    except CareerProcessError as exc:
        if commit:
            db.rollback()
        raise _domain_http_error(exc) from exc
    except Exception:
        if commit:
            db.rollback()
        raise


def _require_user_assertion(source_kind: str) -> None:
    if source_kind != "user_assertion":
        raise HTTPException(
            status_code=422,
            detail=(
                "This authenticated endpoint only accepts user_assertion sources; "
                "Observation, ToolResult, and receipt sources require their trusted "
                "internal handler"
            ),
        )


def _require_direct_action_source(source_kind: str) -> None:
    if source_kind not in {"user_request", "process_event"}:
        raise HTTPException(
            status_code=422,
            detail=(
                "This endpoint only accepts user_request or an owned active "
                "process_event source"
            ),
        )


def _require_direct_transition_source(
    source_kind: str,
    *,
    allow_process_event: bool,
) -> None:
    allowed = {"user_assertion"}
    if allow_process_event:
        allowed.add("process_event")
    if source_kind not in allowed:
        raise HTTPException(
            status_code=422,
            detail=(
                "This endpoint cannot accept an unverified ToolResult, application "
                "result, or CopilotPreference source"
            ),
        )


@router.get("/opportunities", response_model=list[JobOpportunityView])
def get_opportunities(
    include_archived: bool = False,
    limit: int = Query(100, ge=1, le=500),
    offset: int = Query(0, ge=0),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return list_job_opportunities(
        db,
        user_pk=current_user.id,
        include_archived=include_archived,
        limit=limit,
        offset=offset,
    )


@router.post("/opportunities", response_model=JobOpportunityView)
def post_opportunity(
    payload: OpportunityCreate,
    response: Response,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    _require_user_assertion(payload.source_kind)
    if payload.entry_reason == "verified_submission":
        raise HTTPException(
            status_code=422,
            detail=(
                "verified_submission requires a trusted receipt/ToolResult handler; "
                "use user_confirmed_application for a user assertion"
            ),
        )
    admission = _run_domain(
        db,
        lambda: create_job_opportunity(
            db,
            user_pk=current_user.id,
            command=payload,
        ),
        commit=True,
    )
    response.status_code = (
        status.HTTP_201_CREATED if admission.created else status.HTTP_200_OK
    )
    return admission.opportunity


@router.get(
    "/opportunity-merge-candidates",
    response_model=list[OpportunityMergeCandidateView],
)
def get_opportunity_merge_candidates(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return suggest_opportunity_merge_candidates(db, user_pk=current_user.id)


@router.get("/opportunity-merges", response_model=list[OpportunityMergeView])
def get_opportunity_merges(
    include_retracted: bool = False,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return list_opportunity_merges(
        db,
        user_pk=current_user.id,
        include_retracted=include_retracted,
    )


@router.get(
    "/opportunities/{opportunity_id}/jd-snapshots",
    response_model=list[JobDescriptionSnapshotView],
)
def read_job_description_snapshots(
    opportunity_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> list[JobDescriptionSnapshotView]:
    try:
        rows = list_job_description_snapshots(
            db,
            user_pk=current_user.id,
            opportunity_id=opportunity_id,
        )
    except JobDescriptionSnapshotError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return [JobDescriptionSnapshotView.model_validate(row) for row in rows]


@router.post(
    "/opportunities/{opportunity_id}/jd-snapshots",
    response_model=JobDescriptionSnapshotView,
    status_code=status.HTTP_201_CREATED,
)
def create_product_ui_job_description_snapshot(
    opportunity_id: str,
    command: JobDescriptionSnapshotFromProductUI,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> JobDescriptionSnapshotView:
    """Persist an explicit typed product-UI JD submission."""

    try:
        row = create_job_description_snapshot(
            db,
            user_pk=current_user.id,
            opportunity_id=opportunity_id,
            command=command,
        )
        db.commit()
        db.refresh(row)
    except JobDescriptionSnapshotError as exc:
        db.rollback()
        error_name = type(exc).__name__
        status_code = 404 if error_name.endswith("NotFound") else 422
        if error_name.endswith("IdempotencyConflict"):
            status_code = 409
        raise HTTPException(status_code=status_code, detail=str(exc)) from exc
    return JobDescriptionSnapshotView.model_validate(row)


@router.post(
    "/opportunity-merges",
    response_model=OpportunityMergeView,
    status_code=status.HTTP_201_CREATED,
)
def post_opportunity_merge(
    payload: OpportunityMergeCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return _run_domain(
        db,
        lambda: merge_job_opportunities(db, user_pk=current_user.id, command=payload),
        commit=True,
    )


@router.post(
    "/opportunity-merges/{merge_id}/retract",
    response_model=OpportunityMergeView,
)
def post_opportunity_merge_retraction(
    merge_id: str,
    payload: OpportunityMergeRetract,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return _run_domain(
        db,
        lambda: retract_job_opportunity_merge(
            db,
            user_pk=current_user.id,
            merge_id=merge_id,
            command=payload,
        ),
        commit=True,
    )


@router.patch(
    "/opportunities/{opportunity_id}/directions",
    response_model=JobOpportunityView,
)
def patch_opportunity_directions(
    opportunity_id: str,
    payload: OpportunityDirectionsReplace,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    _require_user_assertion(payload.source_kind)
    return _run_domain(
        db,
        lambda: replace_job_opportunity_directions(
            db,
            user_pk=current_user.id,
            opportunity_id=opportunity_id,
            command=payload,
        ),
        commit=True,
    )


@router.get(
    "/opportunities/{opportunity_id}/events",
    response_model=list[ProcessEventView],
)
def get_process_events(
    opportunity_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return _run_domain(
        db,
        lambda: list_process_events(
            db,
            user_pk=current_user.id,
            opportunity_id=opportunity_id,
        ),
    )


@router.post(
    "/opportunities/{opportunity_id}/events",
    response_model=ProcessEventView,
    status_code=status.HTTP_201_CREATED,
)
def post_confirmed_process_event(
    opportunity_id: str,
    payload: ProcessEventAppend,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    _require_user_assertion(payload.source_kind)
    return _run_domain(
        db,
        lambda: append_confirmed_process_event(
            db,
            user_pk=current_user.id,
            opportunity_id=opportunity_id,
            command=payload,
        ),
        commit=True,
    )


@router.post(
    "/opportunities/{opportunity_id}/events/{event_id}/corrections",
    response_model=ProcessEventView,
    status_code=status.HTTP_201_CREATED,
)
def post_process_event_correction(
    opportunity_id: str,
    event_id: str,
    payload: ProcessEventCorrection,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    _require_user_assertion(payload.source_kind)
    return _run_domain(
        db,
        lambda: correct_process_event(
            db,
            user_pk=current_user.id,
            opportunity_id=opportunity_id,
            target_event_id=event_id,
            command=payload,
        ),
        commit=True,
    )


@router.get("/next-actions", response_model=list[NextActionView])
def get_next_actions(
    statuses: list[NextActionStatus] | None = Query(default=None),
    limit: int = Query(200, ge=1, le=500),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return list_next_actions(
        db,
        user_pk=current_user.id,
        statuses=set(statuses) if statuses else None,
        limit=limit,
    )


@router.post(
    "/next-actions",
    response_model=NextActionView,
    status_code=status.HTTP_201_CREATED,
)
def post_next_action(
    payload: NextActionCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    _require_direct_action_source(payload.source_kind)
    return _run_domain(
        db,
        lambda: create_next_action(
            db,
            user_pk=current_user.id,
            command=payload,
        ),
        commit=True,
    )


@router.put("/next-actions/{action_id}", response_model=NextActionView)
def put_next_action(
    action_id: str,
    payload: NextActionEdit,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return _run_domain(
        db,
        lambda: edit_next_action(
            db,
            user_pk=current_user.id,
            action_id=action_id,
            command=payload,
        ),
        commit=True,
    )


@router.post("/next-actions/{action_id}/plan", response_model=NextActionView)
def post_next_action_plan(
    action_id: str,
    payload: NextActionTransition,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    _require_direct_transition_source(
        payload.source_kind,
        allow_process_event=False,
    )
    return _run_domain(
        db,
        lambda: plan_next_action(
            db,
            user_pk=current_user.id,
            action_id=action_id,
            transition=payload,
        ),
        commit=True,
    )


@router.post("/next-actions/{action_id}/complete", response_model=NextActionView)
def post_next_action_complete(
    action_id: str,
    payload: NextActionTransition,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    _require_direct_transition_source(
        payload.source_kind,
        allow_process_event=True,
    )
    return _run_domain(
        db,
        lambda: complete_next_action(
            db,
            user_pk=current_user.id,
            action_id=action_id,
            transition=payload,
        ),
        commit=True,
    )


@router.post("/next-actions/{action_id}/close", response_model=NextActionView)
def post_next_action_close(
    action_id: str,
    payload: NextActionClose,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    _require_direct_transition_source(
        payload.source_kind,
        allow_process_event=True,
    )
    return _run_domain(
        db,
        lambda: close_next_action(
            db,
            user_pk=current_user.id,
            action_id=action_id,
            transition=payload,
        ),
        commit=True,
    )


__all__ = ["router"]
